src/rebuild/srv.c

```diff
@@ -1875,6 +1875,20 @@ rebuild_tgt_prepare(crt_rpc_t *rpc, struct rebuild_tgt_pool_tracker **p_rpt)
 		}
 	}
 
+	if (pool->sp_iv_ns) {
+		/* XXX snapshot will not be allowed to be changed during
+		 * rebuild, let's invalidate local snapshot cache before
+		 * rebuild, so to make sure rebuild will use the updated
+		 * snapshot during rebuild fetch, otherwise it may cause
+		 * corruption.
+		 */
+		rc = cont_iv_snapshot_invalidate(pool->sp_iv_ns,
+						 CRT_IV_SHORTCUT_NONE,
+						 CRT_IV_SYNC_NONE);
```

Comment:
I'm afraid that this simple IV invalidation won't help much here.

Just like taking snapshot, I think we probably need a HLC sync before rebuild, and make the rebuild to only rebuild data with epoch smaller than rebuild HLC, then we only need to guarantee that 'fetch snapshots info HLC A' > 'rebuild HLC' to avoid race of rebuild vs. snapshot creation.

Reply:
Sorry, I am not sure I follow here. For this logical view rebuild,   I think we need rebuild from oldest snapshot to the most recent snapshot, then rebuild by DAOS max epoch.

What do you mean HLC sync?  I assume you mean mark the "start rebuild HLC" over all servers, then snapshot does not touch the HLC below that?  

But how to deal with snapshot deletion, and also not sure how this "HLC sync" will deal with daos max epoch rebuild. 

Oh, I probably miss sth, could you please elaborate a bit?

Reply:
Yes, that's exactly what I'm talking about.

Do we need to rebuild MAX epoch? I think we need only to rebuild data under certain epoch. The data above certain epoch will be written to rebuild target by client, no?

Snapshot deletion won't introduce trouble, if snapshot is deleted during rebuild, rebuild will just rebuild some unnecessary data which will be aggregated on next aggregation.

Reply:
I think we need further discussion for stable view. I will do it in the following patches.